src/planner/binder/query_node/bind_setop_node.cpp

```diff
@@ -71,47 +95,166 @@ unique_ptr<BoundQueryNode> Binder::BindNode(SetOperationNode &statement) {
 	result->left_binder = Binder::CreateBinder(context, this);
 	result->left_binder->can_contain_nulls = true;
 	result->left = result->left_binder->BindNode(*statement.left);
-
 	result->right_binder = Binder::CreateBinder(context, this);
 	result->right_binder->can_contain_nulls = true;
 	result->right = result->right_binder->BindNode(*statement.right);
 
-	if (!statement.modifiers.empty()) {
-		// handle the ORDER BY/DISTINCT clauses
-
-		// we recursively visit the children of this node to extract aliases and expressions that can be referenced in
-		// the ORDER BY
-		case_insensitive_map_t<idx_t> alias_map;
-		expression_map_t<idx_t> expression_map;
-		GatherAliases(*result, alias_map, expression_map);
-
-		// now we perform the actual resolution of the ORDER BY/DISTINCT expressions
-		OrderBinder order_binder({result->left_binder.get(), result->right_binder.get()}, result->setop_index,
-		                         alias_map, expression_map, statement.left->GetSelectList().size());
-		BindModifiers(order_binder, statement, *result);
-	}
-
 	result->names = result->left->names;
 
 	// move the correlated expressions from the child binders to this binder
 	MoveCorrelatedExpressions(*result->left_binder);
 	MoveCorrelatedExpressions(*result->right_binder);
 
 	// now both sides have been bound we can resolve types
-	if (result->left->types.size() != result->right->types.size()) {
+	if (result->setop_type != SetOperationType::UNION_BY_NAME &&
+	    result->left->types.size() != result->right->types.size()) {
 		throw BinderException("Set operations can only apply to expressions with the "
 		                      "same number of result columns");
 	}
 
-	// figure out the types of the setop result by picking the max of both
-	for (idx_t i = 0; i < result->left->types.size(); i++) {
-		auto result_type = LogicalType::MaxLogicalType(result->left->types[i], result->right->types[i]);
-		if (!can_contain_nulls) {
-			if (ExpressionBinder::ContainsNullType(result_type)) {
-				result_type = ExpressionBinder::ExchangeNullType(result_type);
+	if (result->setop_type == SetOperationType::UNION_BY_NAME) {
+		case_insensitive_map_t<idx_t> left_names_map;
```

Comment:
This function is getting very long, could we try moving this into separate functions?

Reply:
I just move this code segment to another big function and add some short comment.